release/apparmor.go

```diff
@@ -66,61 +67,35 @@ func (level AppArmorLevelType) String() string {
 	return fmt.Sprintf("AppArmorLevelType:%d", level)
 }
 
-var (
-	// appArmorLevel contains the assessment of the "level" of apparmor support.
-	appArmorLevel = UnknownAppArmor
-	// appArmorSummary contains a human readable description of the assessment.
-	appArmorSummary string
-	// appArmorKernelFeatures contains a list of kernel features that are supported.
-	// If the value is nil then the features were not probed yet.
-	appArmorKernelFeatures []string
-	// appArmorKernelError contains an error, if any, encountered when
-	// discovering available kernel features.
-	appArmorKernelError error
-	// appArmorParserFeatures contains a list of parser features that are supported.
-	// If the value is nil then the features were not probed yet.
-	appArmorParserFeatures []string
-	// appArmorParserError contains an error, if any, encountered when
-	// discovering available parser features.
-	appArmorParserError error
-)
+// appArmorAssessment represents what is supported AppArmor-wise by the system.
+var appArmorAssessment = &appArmorAssess{appArmorProber: &appArmorProbe{}}
 
 // AppArmorLevel quantifies how well apparmor is supported on the current
 // kernel. The computation is costly to perform. The result is cached internally.
 func AppArmorLevel() AppArmorLevelType {
-	if appArmorLevel == UnknownAppArmor {
-		assessAppArmor()
-	}
-	return appArmorLevel
+	appArmorAssessment.assess()
+	return appArmorAssessment.level
 }
 
 // AppArmorSummary describes how well apparmor is supported on the current
 // kernel. The computation is costly to perform. The result is cached
 // internally.
 func AppArmorSummary() string {
-	if appArmorLevel == UnknownAppArmor {
-		assessAppArmor()
-	}
-	return appArmorSummary
+	appArmorAssessment.assess()
+	return appArmorAssessment.summary
 }
 
 // AppArmorKernelFeatures returns a sorted list of apparmor features like
 // []string{"dbus", "network"}. The result is cached internally.
 func AppArmorKernelFeatures() ([]string, error) {
-	if appArmorKernelFeatures == nil {
-		appArmorKernelFeatures, appArmorKernelError = probeAppArmorKernelFeatures()
-	}
-	return appArmorKernelFeatures, appArmorKernelError
+	return appArmorAssessment.KernelFeatures()
```

Comment:
Mostly curiuous - is there a reason we use slightly different pattern here and above? I mean explicit call to `assess()`, rather than having `Level()` and `Summary()` methods that to `sync.Once` internally?

Reply:
These ones are used both internally and externally and also participate in the interface. Summary and Level are just exposed externally so having another layer of methods didn't seem interesting just for the symmetry.